python/mspasspy/db/database.py

```diff
@@ -308,39 +341,93 @@ def save_data(self, mspass_object, storage_mode='gridfs', dfile=None, dir=None,
 
-            # 2. save data
-            wf_collection = save_schema.collection('_id') if not collection else collection
-            col = self[wf_collection]
-            object_doc = col.find_one({'_id': mspass_object['_id']})
-            filter_ = {'_id': mspass_object['_id']}
-            update_dict = {'storage_mode': storage_mode}
-
-            if storage_mode == "file":
-                foff = self._save_data_to_dfile(mspass_object, dir, dfile)
-                update_dict['dir'] = dir
-                update_dict['dfile'] = dfile
-                update_dict['foff'] = foff
-            elif storage_mode == "gridfs":
-                old_gridfs_id = None if 'gridfs_id' not in object_doc else object_doc['gridfs_id']
-                gridfs_id = self._save_data_to_gridfs(mspass_object, old_gridfs_id)
-                update_dict['gridfs_id'] = gridfs_id
-            #TODO will support url mode later 
-            #elif storage_mode == "url":
-            #    pass
-
-            col.update_one(filter_, {'$set': update_dict})
-        else:
+            if mspass_object.live:
+                # 2. save data
+                wf_collection = save_schema.collection('_id') if not collection else collection
+                col = self[wf_collection]
+                object_doc = col.find_one({'_id': mspass_object['_id']})
+                filter_ = {'_id': mspass_object['_id']}
+                update_dict = {'storage_mode': storage_mode}
+
+                if storage_mode == "file":
+                    foff = self._save_data_to_dfile(mspass_object, dir, dfile)
+                    update_dict['dir'] = dir
+                    update_dict['dfile'] = dfile
+                    update_dict['foff'] = foff
+                elif storage_mode == "gridfs":
+                    old_gridfs_id = None if 'gridfs_id' not in object_doc else object_doc['gridfs_id']
+                    gridfs_id = self._save_data_to_gridfs(mspass_object, old_gridfs_id)
+                    update_dict['gridfs_id'] = gridfs_id
+                #TODO will support url mode later 
+                #elif storage_mode == "url":
+                #    pass
+                col.update_one(filter_, {'$set': update_dict})
+            else:
+                save_metadata_fail = True
+        
+        if not mspass_object.live:
             # FIXME: we could have recorded the full stack here, but need to revise the logger object
             # to make it more powerful for Python logging.
             mspass_object.elog.log_verbose(
                 sys._getframe().f_code.co_name, "Skipped saving dead object")
             self._save_elog(mspass_object)
+        
+        return 1 if save_metadata_fail else 0
+
+    # clean the collection fixing any type errors and removing any aliases using the schema currently defined for self
+    def clean(self, collection, log_id_keys=[], print=False, query=None):
```

Comment:
This is not covered by test.

Reply:
implement another `clean` method for document and change this one to `clean_collection` and wrap around `clean`